okhttp/src/commonJvmAndroid/kotlin/okhttp3/Request.kt

```diff
@@ -82,8 +82,8 @@ class Request internal constructor(
 
   init {
     val connectionHeader = headers["Connection"]
-    require(body == null || !"upgrade".equals(connectionHeader, ignoreCase = true)) {
-      "expected a null request body with 'Connection: upgrade'"
+    require(body == null || !"upgrade".equals(connectionHeader, ignoreCase = true) || body.contentLength() <= 0) {
```

Comment:
Or even `body.contentLength() == 0`

Reply:
It’s gotta be `body.contentLength() == 0`. If we permit `body.contentLength() == -1` that is not an empty body.